.travis.yml

```diff
@@ -2,47 +2,85 @@ sudo: false
 language: ruby
 script: bin/test
 install: bin/bootstrap --path vendor/bundle --without development debug
-rvm:
-  - 1.9.3
-  - 2.0.0
-  - 2.1.10
-  - 2.2.7
-  - 2.3.4
-  - 2.4.1
-  - ruby-head
-  - jruby
-  - jruby-20mode
-  - jruby-21mode
-  - jruby-9.1.12.0
 matrix:
   include:
+    - rvm: ruby-head
+      os: linux
+    - rvm: ruby-head
+      os: osx
+    - rvm: 2.4.1
+      os: linux
+      # TravisCI OSX doesn't work with Ruby 2.4.1 yet (https://github.com/travis-ci/travis-ci/issues/7848)
+    - rvm: 2.4.0
+      os: osx
+    - rvm: 2.3.4
+      os: linux
+    - rvm: 2.3.4
+      os: osx
+    - rvm: 2.2
+      os: linux
+    - rvm: 2.2
+      os: osx
+    - rvm: 2.1
+      os: linux
+    - rvm: 2.1
+      os: osx
+    - rvm: 2.0.0
+      os: linux
+    - rvm: 2.0.0
+      os: osx
+    - rvm: 1.9.3
+      os: linux
+    - rvm: 1.9.3
+      os: osx
+    - rvm: jruby
+      os: linux
+      env:
+        # Travis by default also have "-Dcext.enabled=false" set in
+        # JRUBY_OPTS, but JRuby 9 does not support C extensions at all
+        # so it issues warning that will mess up the sterr checks.
+        - JRUBY_OPTS="--server -Xcompile.invokedynamic=false"
+        # Somehow a "ASCII-8BIT to UTF-8 conversion error" appears for
+        # JRuby 9
+        - LC_ALL=en_US.UTF-8
+        - LANG=en_US.UTF-8
+        - LANGUAGE=en_US.UTF-8
     - rvm: jruby-9.1.12.0
-      env: JRUBY_OPTS='--dev'
-    - rvm: jruby-9.1.12.0-20mode
-      env: JRUBY_OPTS='--dev'
-    - rvm: jruby-9.1.12.0-21mode
-      env: JRUBY_OPTS='--dev'
+      os: linux
+      env:
+        # Travis by default also have "-Dcext.enabled=false" set in
+        # JRUBY_OPTS, but JRuby 9 does not support C extensions at all
+        # so it issues warning that will mess up the sterr checks.
+        - JRUBY_OPTS="--server -Xcompile.invokedynamic=false"
+        # Somehow a "ASCII-8BIT to UTF-8 conversion error" appears for
+        # JRuby 9
+        - LC_ALL=en_US.UTF-8
+        - LANG=en_US.UTF-8
+        - LANGUAGE=en_US.UTF-8
+    - rvm: jruby-9.1.12.0
+      os: osx
+      env:
+        # Travis by default also have "-Dcext.enabled=false" set in
+        # JRUBY_OPTS, but JRuby 9 does not support C extensions at all
+        # so it issues warning that will mess up the sterr checks.
+        - JRUBY_OPTS="--server -Xcompile.invokedynamic=false"
```

Comment:
Can we move setting of these variables to a global section to avoid duplication?